// github-data-cat/lib/src/structrue/forks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const forkCountPerPage = 20;
function parseFork(forkInfo) {
    return {
        login: forkInfo.owner.login,
        time: forkInfo.createdAt
    };
}
async function getForks(client, owner, name, updatedAfter) {
    let forks = [];
    let forksInfo;
    let pageInfo;
    do {
        forksInfo = await client.query(getMoreIssueSql, {
            owner,
            name,
            forkCount: forkCountPerPage,
            cursor: pageInfo ? pageInfo.endCursor : null
        });
        if (!forksInfo || !forksInfo.repository || !forksInfo.repository.forks || forksInfo.repository.forks.nodes.length === 0)
            break;
        if (updatedAfter && new Date(forksInfo.repository.forks.nodes[forksInfo.repository.forks.nodes.length - 1].createdAt) <= updatedAfter) {
            forks = forks.concat(forksInfo.repository.forks.nodes.filter(f => f.owner && f.owner.login && new Date(f.createdAt) >= updatedAfter).map(parseFork));
            break;
        }
        forks = forks.concat(forksInfo.repository.forks.nodes.filter(f => f.owner).map(parseFork));
        pageInfo = forksInfo.repository.forks.pageInfo;
    } while (pageInfo.hasNextPage);
    return forks;
}
exports.getForks = getForks;
const getMoreIssueSql = `
query getMoreForks($owner: String!, $name: String!, $forkCount: Int, $cursor: String) {
    rateLimit {
        resetAt
        remaining
        cost
    }
    repository(owner: $owner, name: $name) {
        forks(first: $forkCount, orderBy: {field: UPDATED_AT, direction: DESC}, after: $cursor) {
            pageInfo {
                hasNextPage
                endCursor
            }
            nodes {
                owner {
                    login
                }
                createdAt
            }
        }
    }
}
`;
