// github-data-cat/lib/src/structrue/pull-requests.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const pullRequestsPerPage = 5;
const commentCount = 100;
const labelCount = 10;
const commitCount = 100;
const reviewCount = 100;
const reviewCommentCount = 100;
function parsePullRequest(pr) {
    let reivewComments = [];
    pr.reviews.nodes.forEach(r => {
        r.comments.nodes.filter(c => c.author).forEach(c => {
            reivewComments.push({
                id: c.id,
                url: c.url,
                login: c.author.login,
                body: c.body,
                createdAt: c.createdAt
            });
        });
    });
    let additions = 0;
    let deletions = 0;
    pr.commits.nodes.forEach(ci => {
        additions += ci.commit.additions;
        deletions += ci.commit.deletions;
    });
    return {
        id: pr.id,
        author: pr.author.login,
        number: pr.number,
        createdAt: pr.createdAt,
        updatedAt: pr.updatedAt,
        closedAt: pr.closedAt,
        mergedAt: pr.mergedAt,
        title: pr.title,
        body: pr.body,
        labels: pr.labels.nodes.map(l => l.name),
        comments: pr.comments.nodes.filter(c => c.author).map(c => {
            return {
                id: c.id,
                login: c.author.login,
                body: c.body,
                url: c.url,
                createdAt: c.createdAt
            };
        }),
        reviewComments: reivewComments,
        additions,
        deletions
    };
}
async function getPullRequests(client, owner, name, updatedAfter) {
    let prsInfo;
    let prs = [];
    let pageInfo;
    do {
        prsInfo = await client.query(getMoreIssueSql, {
            owner,
            name,
            prCount: pullRequestsPerPage,
            cursor: pageInfo ? pageInfo.endCursor : null,
            labelCount,
            commentCount,
            reviewCount,
            commitCount,
            reviewCommentCount
        });
        if (!prsInfo || !prsInfo.repository || !prsInfo.repository.pullRequests || prsInfo.repository.pullRequests.nodes.length === 0)
            break;
        if (updatedAfter && new Date(prsInfo.repository.pullRequests.nodes[prsInfo.repository.pullRequests.nodes.length - 1].updatedAt) <= updatedAfter) {
            prs = prs.concat(prsInfo.repository.pullRequests.nodes.filter(pr => pr.author && new Date(pr.updatedAt) >= updatedAfter).map(parsePullRequest));
            break;
        }
        prs = prs.concat(prsInfo.repository.pullRequests.nodes.filter(pr => pr.author).map(parsePullRequest));
        pageInfo = prsInfo.repository.pullRequests.pageInfo;
    } while (pageInfo.hasNextPage);
    return prs;
}
exports.getPullRequests = getPullRequests;
const getMoreIssueSql = `
query getMorePullRequests($owner: String!, $name: String!, $prCount: Int, $cursor: String, $labelCount: Int, $commentCount: Int, $commitCount: Int, $reviewCount: Int, $reviewCommentCount: Int) {
    rateLimit {
      resetAt
      remaining
      cost
    }
    repository(owner: $owner, name: $name) {
      pullRequests(first: $prCount, orderBy: {field: UPDATED_AT, direction: DESC}, after: $cursor) {
        pageInfo {
          hasNextPage
          endCursor
        }
        nodes {
          id: databaseId
          title
          body
          number
          createdAt
          updatedAt
          closedAt
          mergedAt
          author {
            login
          }
          labels(first: $labelCount) {
            nodes {
              name
            }
          }
          comments(first: $commentCount) {
            nodes {
              id: databaseId
              body
              url
              author {
                login
              }
              createdAt
            }
          }
          commits(first: $commitCount) {
            nodes {
              commit {
                additions
                deletions
              }
            }
          }
          reviews(first: $reviewCount) {
            nodes {
              comments(first: $reviewCommentCount) {
                nodes {
                  id: databaseId
                  url
                  author {
                    login
                  }
                  body
                  createdAt
                }
              }
            }
          }
        }
      }
    }
  }
`;
