// github-data-cat/lib/src/structrue/stars.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const starCountPerPage = 20;
function parseStar(starInfo) {
    return {
        login: starInfo.node.login,
        time: starInfo.starredAt
    };
}
async function getStars(client, owner, name, updatedAfter) {
    let stars = [];
    let starInfo;
    let pageInfo;
    do {
        starInfo = await client.query(getStarSql, {
            owner,
            name,
            starCount: starCountPerPage,
            cursor: pageInfo ? pageInfo.endCursor : null
        });
        if (!starInfo || !starInfo.repository || !starInfo.repository.stargazers || starInfo.repository.stargazers.edges.length === 0)
            break;
        if (updatedAfter && new Date(starInfo.repository.stargazers.edges[starInfo.repository.stargazers.edges.length - 1].starredAt) <= updatedAfter) {
            stars = stars.concat(starInfo.repository.stargazers.edges.filter(e => e.node && e.node.login && new Date(e.starredAt) >= updatedAfter).map(parseStar));
            break;
        }
        stars = stars.concat(starInfo.repository.stargazers.edges.filter(e => e.node && e.node.login).map(parseStar));
        pageInfo = starInfo.repository.stargazers.pageInfo;
    } while (pageInfo.hasNextPage);
    return stars;
}
exports.getStars = getStars;
const getStarSql = `
query getMoreStars($owner: String!, $name: String!, $starCount: Int, $cursor: String) {
    rateLimit {
        resetAt
        remaining
        cost
    }
    repository(owner: $owner, name: $name) {
        stargazers(first: $starCount, orderBy: {field: STARRED_AT, direction: DESC}, after: $cursor) {
            pageInfo {
                hasNextPage
                endCursor
            }
            edges {
                starredAt
                node {
                    login
                }
            }
        }
    }
}
`;
