// github-data-cat/lib/src/data_cat.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Logger = require("bunyan");
const github_graphql_v4_client_1 = require("github-graphql-v4-client");
const organizations_1 = require("./structrue/organizations");
const repos_1 = require("./structrue/repos");
const stars_1 = require("./structrue/stars");
const forks_1 = require("./structrue/forks");
const issues_1 = require("./structrue/issues");
const pull_requests_1 = require("./structrue/pull-requests");
const contributors_1 = require("./structrue/contributors");
class DataCat {
    constructor(options) {
        this.logger = Logger.createLogger({
            name: "GitHub-GraphQL-Fetcher",
            level: Logger.ERROR
        });
        this.maxConcurrentReqNumber = 10;
        this.filterStatusCode = [400, 401, 403, 404];
        this.maxRetryTimes = 10;
        this.org = {
            repos: (login, updatedAfter) => organizations_1.getRepos(login, this.client, updatedAfter)
        };
        this.repo = {
            info: (owner, name) => repos_1.getRepo(owner, name, this.client),
            stars: (owner, name, updatedAfter) => stars_1.getStars(this.client, owner, name, updatedAfter),
            forks: (owner, name, updatedAfter) => forks_1.getForks(this.client, owner, name, updatedAfter),
            issues: (owner, name, updatedAfter) => issues_1.getIssues(this.client, owner, name, updatedAfter),
            pulls: (owner, name, updatedAfter) => pull_requests_1.getPullRequests(this.client, owner, name, updatedAfter),
            contributors: (owner, name, branch, commitLimit) => contributors_1.getContributors(this.client, owner, name, branch, commitLimit),
            full: async (owner, name, params, updatedAfter) => {
                let repo = await this.repo.info(owner, name);
                let [stars, forks, issues, pulls, contributors] = await Promise.all([
                    (params && params.stars && repo.starCount > 0) ? this.repo.stars(owner, name, updatedAfter) : [],
                    (params && params.forks && repo.forkCount > 0) ? this.repo.forks(owner, name, updatedAfter) : [],
                    (params && params.issues) ? this.repo.issues(owner, name, updatedAfter) : [],
                    (params && params.pulls) ? this.repo.pulls(owner, name, updatedAfter) : [],
                    (params && params.contributors) ?
                        // if get contributors, then only get these after repo created
                        this.repo.contributors(owner, name, repo.defaultBranchName) : []
                ]);
                repo.stars = stars;
                repo.forks = forks;
                repo.issues = issues;
                repo.pulls = pulls;
                repo.contributors = contributors;
                return repo;
            }
        };
        this.inited = false;
        if (options.tokens.length === 0) {
            throw new Error('At least one token needed.');
        }
        this.tokens = options.tokens;
        if (options.logger) {
            this.logger = options.logger;
        }
        if (options.filterStatusCode) {
            this.filterStatusCode = options.filterStatusCode;
        }
        if (options.maxRetryTimes) {
            this.maxRetryTimes = options.maxRetryTimes;
        }
        if (options.maxConcurrentReqNumber) {
            this.maxConcurrentReqNumber = options.maxConcurrentReqNumber;
        }
    }
    async init() {
        this.client = new github_graphql_v4_client_1.GitHubClient({
            tokens: this.tokens,
            logger: this.logger,
            maxConcurrentReqNumber: this.maxConcurrentReqNumber,
            filterStatusCode: this.filterStatusCode,
            maxRetryTimes: this.maxRetryTimes
        });
        await this.client.init();
        this.inited = true;
    }
}
exports.DataCat = DataCat;
