// github-graphql-v4-client/lib/index.js
"use strict";
/* Copyright 2019 ZhaoShengyu
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at

 *    http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("@octokit/graphql");
const Logger = require("bunyan");
const waitFor = require("p-wait-for");
class GitHubClient {
    constructor(options) {
        this.tokens = [];
        this.maxConcurrentReqNumber = 10;
        this.concurrentReqNumber = 0;
        this.logger = Logger.createLogger({
            level: Logger.INFO,
            name: 'GitHub-GraphQL-Client'
        });
        this.getConnectionRetryInterval = 10000;
        this.filterStatusCode = [400, 401, 403, 404];
        this.requestCostPrediction = 15;
        this.maxRetryTimes = 10;
        if (options.tokens.length === 0) {
            throw new Error('At least one token needed.');
        }
        if (options.logger) {
            this.logger = options.logger;
        }
        if (options.filterStatusCode) {
            this.filterStatusCode = options.filterStatusCode;
        }
        if (options.maxRetryTimes) {
            this.maxRetryTimes = options.maxRetryTimes;
        }
        if (options.maxConcurrentReqNumber) {
            this.maxConcurrentReqNumber = options.maxConcurrentReqNumber;
        }
        this.tokens = options.tokens.map((t) => {
            return {
                token: t,
                ratelimitRemaining: -1,
                ratelimitReset: -1
            };
        });
        this.inited = false;
    }
    async init() {
        await Promise.all(this.tokens.map((token) => this.initToken(token)));
        this.logger.info(`Token inited done, tokens = ${JSON.stringify(this.tokens)}`);
        this.inited = true;
    }
    async query(q, p) {
        if (!this.inited) {
            throw new Error('Client not inited yet! Call `await client.init()` to init.');
        }
        // add rate limit
        const firstBraceIndex = q.indexOf('{') + 1;
        q = q.substr(0, firstBraceIndex) + rateLimitQueryStr + q.substr(firstBraceIndex);
        return this.internalQuery(q, p, 0);
    }
    resetToken(token) {
        if (this.hasSufficientRemaing(token)) {
            return;
        }
        // add 1s to ensure reset on server side
        let resetTime = new Date(token.ratelimitReset).getTime() - new Date().getTime() + 1000;
        if (resetTime < 0) {
            this.logger.error(`Something wrong with rate limit maintain.`);
            resetTime = 10 * 60 * 1000;
        }
        setTimeout(() => {
            this.initToken(token);
        }, resetTime);
    }
    async initToken(token) {
        const response = await graphql_1.graphql(rateLimitQuerySql, { headers: { authorization: `token ${token.token}` } });
        token.ratelimitRemaining = response.rateLimit.remaining;
        token.ratelimitReset = response.rateLimit.resetAt;
        this.resetToken(token);
    }
    // get a valid token
    async getToken() {
        let token;
        await waitFor(() => {
            if (this.concurrentReqNumber >= this.maxConcurrentReqNumber) {
                return false;
            }
            const availableTokens = this.tokens.filter((t) => this.hasSufficientRemaing(t));
            if (availableTokens.length === 0) {
                this.logger.warn(`No avialable token found for now, will try later`);
                return false;
            }
            this.concurrentReqNumber += 1;
            token = availableTokens[Math.floor(Math.random() * availableTokens.length)];
            return true;
        }, {
            interval: this.getConnectionRetryInterval
        });
        return token;
    }
    hasSufficientRemaing(token) {
        return token.ratelimitRemaining > this.requestCostPrediction * this.maxConcurrentReqNumber;
    }
    // query function
    async internalQuery(q, p, retryCount) {
        const token = await this.getToken();
        try {
            // set auth token
            Object.assign(p, { headers: { authorization: `token ${token.token}` } });
            const res = (await graphql_1.graphql(q, p));
            const rateLimitRes = res;
            this.concurrentReqNumber--;
            if (!rateLimitRes.rateLimit) {
                this.logger.error(`No rate limit returned for query = ${q}, param = ${JSON.stringify(p)}`);
                process.exit(1);
            }
            token.ratelimitRemaining = rateLimitRes.rateLimit.remaining;
            token.ratelimitReset = rateLimitRes.rateLimit.resetAt;
            this.resetToken(token);
            return res;
        }
        catch (e) {
            this.logger.error(`Error happened, e = ${JSON.stringify(e)}`);
            this.concurrentReqNumber--;
            let apiRateLimitExceeded = false;
            const response = e;
            if (response.errors) {
                // log error if exists
                if (response.errors.find((e) => e.message.includes('API rate limit exceeded'))) {
                    // rate limit exceeded
                    this.logger.warn(`Token API rate limit exceeded, token = ${JSON.stringify(token)}`);
                    apiRateLimitExceeded = true;
                    this.resetToken(token);
                }
                else if (response.errors.find((e) => e.type && e.type === 'NOT_FOUND')) {
                    // not found, maybe deleted
                    return null;
                }
            }
            if (retryCount >= this.maxRetryTimes) {
                // retry times exceed the max retry times, return the data no matter there is any
                this.logger.warn(`Retry time exceed max retry times. query = ${q}`);
                return response.data;
            }
            if (apiRateLimitExceeded
                || !response.status
                || (response.status >= 400 && !this.filterStatusCode.includes(response.status))) {
                // api rate limit exceeded
                // no status field
                // status >= 400
                return this.internalQuery(q, p, retryCount + 1);
            }
            else if (response.data) {
                // other status code, return data if exists
                return response.data;
            }
        }
        return null;
    }
}
exports.GitHubClient = GitHubClient;
const rateLimitQueryStr = `
rateLimit {
    resetAt
    remaining
}
`;
const rateLimitQuerySql = `
query {
    ${rateLimitQueryStr}
}
`;
