// open-md-checker/lib/index.js
#!/usr/bin/env node
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const fs_1 = require("fs");
const path_1 = require("path");
const default_requires_1 = tslib_1.__importDefault(require("./default_requires"));
// tslint:disable-next-line: no-var-requires
const deglob = require('deglob');
// tslint:disable-next-line: no-var-requires
const gutil = require('gulp-util');
const moduleName = 'open-md-checker';
function loadConfig() {
    const defaultConfig = {
        patterns: ['**/*.md'],
        options: {
            useGitIgnore: true,
            ignore: ['node_modules/**/*'],
        },
    };
    const configFile = process.env.MD_LINT_CONFIG_PATH;
    if (!configFile) {
        gutil.log(moduleName, gutil.colors.yellow('Not set env MD_LINT_CONFIG_PATH, use default config'));
        return defaultConfig;
    }
    if (!fs_1.existsSync(configFile)) {
        gutil.log(moduleName, gutil.colors.yellow(`Can not found config file ${configFile}, use default config`));
        return defaultConfig;
    }
    const configs = fs_1.readFileSync(configFile).toString();
    if (configs) {
        const json = JSON.parse(configs);
        if (json[moduleName])
            return json[moduleName];
    }
    gutil.log(moduleName, gutil.colors.yellow(`Load config from ${configFile} error, use default config`));
    return defaultConfig;
}
function getRemarkInstance() {
    const remark = require('remark');
    let md_lint_requires = default_requires_1.default();
    // load MarkdownLint
    if (config.requires) {
        if (!config.requires.startsWith('/')) {
            config.requires = path_1.join(process.cwd(), config.requires);
        }
        if (!fs_1.existsSync(config.requires)) {
            gutil.log(moduleName, gutil.colors.yellow(`Can not found md-lint-requires file ${config.requires}, use default config`));
        }
        else {
            gutil.log(moduleName, gutil.colors.yellow(`Load md-lint-requires from ${config.requires}`));
            md_lint_requires = require(config.requires);
        }
    }
    return remark().use(md_lint_requires);
}
function doCheck(files) {
    const report = require('vfile-reporter');
    const remarkInstance = getRemarkInstance();
    files.forEach(filePath => {
        const mdContent = fs_1.readFileSync(filePath, 'utf8');
        if (mdContent) {
            remarkInstance.process(mdContent, (err, file) => {
                const res = report(err || file);
                console.log(filePath);
                console.log(res);
                console.log();
                if (res !== 'no issues found')
                    process.exitCode = -1;
            });
        }
    });
}
const config = loadConfig();
deglob(config.patterns, config.options, (err, files) => {
    if (err) {
        gutil.log(moduleName, gutil.colors.red(err));
        process.exit(-1);
    }
    doCheck(files);
});
